libbpf-tools/bashreadline.c

```diff
@@ -0,0 +1,218 @@
+/* SPDX-License-Identifier: (LGPL-2.1 OR BSD-2-Clause) */
+#include <argp.h>
+#include <stdio.h>
+#include <errno.h>
+#include <signal.h>
+#include <time.h>
+#include <unistd.h>
+#include <ctype.h>
+
+#include <bpf/libbpf.h>
+#include <bpf/bpf.h>
+#include "bashreadline.h"
+#include "bashreadline.skel.h"
+#include "trace_helpers.h"
+#include "uprobe_helpers.h"
+
+#define PERF_BUFFER_PAGES	16
+#define PERF_POLL_TIMEOUT_MS	100
+#define warn(...) fprintf(stderr, __VA_ARGS__)
+
+static volatile sig_atomic_t exiting = 0;
+
+const char *argp_program_version = "bashreadline 1.0";
+const char *argp_program_bug_address =
+	"https://github.com/iovisor/bcc/tree/master/libbpf-tools";
+const char argp_program_doc[] =
+"Print entered bash commands from all running shells.\n"
+"\n"
+"USAGE: bashreadline [-s <path/to/libreadline.so>]\n"
+"\n"
+"EXAMPLES:\n"
+"    bashreadline\n"
+"    bashreadline -s /usr/lib/libreadline.so\n";
+
+static const struct argp_option opts[] = {
+	{ "shared", 's', "PATH", 0, "the location of libreadline.so library" },
+	{ NULL, 'h', NULL, OPTION_HIDDEN, "Show the full help" },
+	{},
+};
+
+static char *libreadline_path = NULL;
+
+static error_t parse_arg(int key, char *arg, struct argp_state *state)
+{
+	switch (key) {
+	case 's':
+		libreadline_path = strdup(arg);
+		if (libreadline_path == NULL)
+			return ARGP_ERR_UNKNOWN;
+		break;
+	case 'h':
+		argp_state_help(state, stderr, ARGP_HELP_STD_HELP);
+		break;
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return 0;
+}
+
+static void handle_event(void *ctx, int cpu, void *data, __u32 data_size)
+{
+	struct str_t *e = data;
+	struct tm *tm;
+	char ts[16];
+	time_t t;
+
+	time(&t);
+	tm = localtime(&t);
+	strftime(ts, sizeof(ts), "%H:%m:%S", tm);
+
+	printf("%-9s %-7d %s\n", ts, e->pid, e->str);
+}
+
+static void handle_lost_events(void *ctx, int cpu, __u64 lost_cnt)
+{
+	warn("lost %llu events on CPU #%d\n", lost_cnt, cpu);
+}
+
+static char *find_readline_so()
+{
+	const char *bash_path = "/bin/bash";
+	FILE *fp;
+	char *line = NULL;
+	char *sep, *path_start, *sp;
+	size_t line_sz;
+	off_t func_off;
+	char *result = NULL;
+
+	func_off = get_elf_func_offset(bash_path, "readline");
+	if (func_off >= 0)
+		return strdup(bash_path);
+
+	/*
+	 * Try to find libreadline.so if the symbol of readline is not
+	 * defined in bash
+	 */
+	fp = popen("ldd /bin/bash", "r");
+	if (fp == NULL)
+		goto cleanup;
+	while (getline(&line, &line_sz, fp) >= 0) {
+		if (strstr(line, "libreadline.so") == NULL)
+			continue;
+		sep = strstr(line, " => ");
+		if (sep == NULL)
+			goto cleanup;
+		path_start = sep + 4;
+		while (isspace(*path_start))
+			path_start++;
+		sp = strstr(path_start, " ");
+		if (sp == NULL)
+			goto cleanup;
```

Comment:
tbh, I'm a bit lost of what kind of string we are parsing here, can you please add a comment with an example? It would make it easier to follow and maybe someone can suggest a more succinct way to go about that